### PR TITLE
[Common] Add dense router output for fused router

### PR DESCRIPTION
**Pull request by harryzhou2000**
The fused router forward kernels currently produce boolean or bitmap sparse routing map format, which may need an additional conversion to `topk_indices` format (in the shape of `[*leading_dims, top_k]`) to be passed to the dispatcher. For example, NCCL EP accepts `topk_indices` as routing map. To avoid needing an extra kernel for routing map conversion, the fused router could directly write into that format.

For NCCL EP, the dense `topk_indices` row is consumed as an order-insensitive selected-expert set. The dense output therefore does not promise score-sorted or expert-sorted order; it preserves the selected experts produced by the chosen top-k kernel path.

## Summary
- Add an optional dense `topk_indices` output path to fused router top-k.
- Avoid materializing the full routing map when callers provide a `[*leading_dims, topk]` index buffer.
- Add dense-index backward using the selected expert indices directly.
- Support `int16`, `int32`, and `int64` dense index buffers.
- Preserve existing BYTEMAP/BITMAP_U8 routing-map paths and p3R radix/naive dispatch behavior.
- Add guard checks for dense index shape/device/dtype, grouped top-k assumptions, routing-map format, and direct score-function API usage.
- Add `int16` support for TE CUDA graph weak-ref tensors.

## Testing
- Built on B200 with:
  `NVTE_BUILD_THREADS_PER_JOB=4 NVTE_CUDA_ARCHS="90;100;103a;120" NVTE_USE_CCACHE=1 pip install --no-build-isolation -e .[test] --verbose`
- Ran:
  `python -m pytest -q tests/pytorch/test_fused_router.py`
- Result:
  `3203 passed, 444 skipped, 3 warnings in 44.17s`

### CODE REVIEW

File: transformer_engine/pytorch/csrc/extensions/router.cpp [hunk 1]
```diff
@@ -87,38 +134,57 @@ std::tuple<at::Tensor, at::Tensor, at::Tensor> fused_topk_with_score_function_fw
...
                                         bool use_pre_softmax, std::optional<float> scaling_factor,
-                                        std::string score_function, int routing_map_format) {
+                                        std::string score_function, bool use_dense_indices,
+                                        int routing_map_format) {
+  check_routing_map_format(routing_map_format);
   TORCH_CHECK(grad_probs.dim() >= 1, "grad_probs must have at least 1 dim");
   TORCH_CHECK(grad_probs.is_contiguous(), "grad_probs must be contiguous");
   TORCH_CHECK(grad_logits.is_contiguous(), "grad_logits must be contiguous");
   auto sizes = grad_probs.sizes();
   int64_t num_experts = sizes.back();
```
[the remark this answers is not in the document]
Comment (harryzhou2000) [reply transformer_engine/pytorch/csrc/extensions/router.cpp:173]: Fixed in `5ad179b4`. Dense-index backward now rejects non-default `routing_map_format` explicitly; non-dense backward still validates BYTEMAP/BITMAP_U8 normally.

File: transformer_engine/pytorch/csrc/extensions/router.cpp [hunk 2]
```diff
@@ -54,6 +93,9 @@ std::tuple<at::Tensor, at::Tensor, at::Tensor> fused_topk_with_score_function_fw
   if (score_function == "sigmoid" || score_function == "sqrtsoftplus") {
     use_pre_softmax = false;  // Pre-softmax only happens at the softmax case
   }
+  if (topk_indices.has_value()) {
+    check_dense_topk_indices(topk_indices.value(), logits, num_tokens, topk);
+  }
```
[the remark this answers is not in the document]
Comment (harryzhou2000) [reply transformer_engine/pytorch/csrc/extensions/router.cpp:100]: Fixed in `5ad179b4`. The C++ wrapper now rejects `topk_indices` together with a non-default `routing_map_format`, so callers cannot silently request BITMAP_U8 while taking the dense-index path.

File: transformer_engine/pytorch/router.py [hunk 3]
```diff
@@ -95,22 +100,35 @@ def forward(
...
+        ctx.tensor_shape = tensor_shape
         ctx.use_pre_softmax = use_pre_softmax
         ctx.topk = topk
         ctx.scaling_factor = scaling_factor
         ctx.score_function = score_function
         ctx.routing_map_format = routing_map_format
-        return probs, routing_map
+        ctx.logits_dtype = logits.dtype
+        ctx.use_dense_indices = topk_indices is not None
+        return probs, routing_output
```
[the remark this answers is not in the document]
Comment (harryzhou2000) [reply transformer_engine/pytorch/router.py]: Fixed in `49c65536`. The Python wrapper no longer flattens `logits`; C++ preserves the original leading dimensions for sparse routing maps, and dense `topk_indices` now requires the same leading dimensions with trailing `topk`.

**Review by denera — APPROVED** [review]
LGTM, looks like all the issues are addressed already. CI is clean except for unrelated known JAX issue (reran license failure and it passed). Thanks!
